accepted/build/build-publish-tweaks.md

```diff
@@ -0,0 +1,28 @@
+# Build and Publish tweaks in the .NET Core SDK
+
+In .NET Core 3, we are making Build output more similar to the Publish output.  In .NET Core 2.x and prior, dependencies from NuGet packages are not copied into the Build output.  At runtime, they are loaded from the NuGet package cache.  This meant that the Build output was not portable to other machines.
+
+For .NET Core 3, the Build output includes an app's dependencies.  The Build and Publish output from the base .NET SDK should be the same.  Additional operations that are not part of the inner-loop development cycle can be part of Publish, such as copying content for Web apps, trimming unused dependencies, or "linking" the app to create a single file executable.
+
+This document proposes some changes to the SDK behavior that will help us unify the Build and Publish logic.  Some of these could be breaking changes, but we believe the impact will be low.
+
+## `Publish="False"` on `PackageReference` should apply to Build
+
+If a `PackageReference` has `Publish="False"` metadata, then assets from that package will not be included in the Publish output.  In current previews of .NET Core 3, those assets will be included in the Build output but not the Publish output.  However, when targeting previous versions of .NET Core, no assets from NuGet packages would be included in the Build output.  Also, we are not aware of a scenario where you would want to specify that assets should not be in the Publish output, but would still want them in the Build output.
```

Comment:
> Also, we are not aware of a scenario where you would want to specify that assets should not be in the Publish output, but would still want them in the Build output.

My understanding of the original scenario here was to do with an "original CLI Tool" feature. In that feature/scenario, the tool's dependency graph and the project's dependency graph would be unified, and the tool would run in the context of your project. That way the tool could do things like load types/execute code/etc. I believe that was how the original `dotnet test` (pre-vstest implementation) worked.

So with that scenario, you'd want to run a CLI Tool against your build-time assets, and some of those dependencies (like the tool itself) weren't needed at publish time.

I _hope_ the above CLI Tooling feature/functionality is no longer in use. But that is my understanding of the driving scenario here to have a "build-time only" dependencies.

Reply:
> I hope the above CLI Tooling feature/functionality is no longer in use.

How is this functionality triggered. Is there an option on the tool reference or was this something that predates 1.0 even? We're making other changes to limit what CLI tools can do. For example, maximum TFM of 2.2 due to some design flaws. CLI tools are deprecated vs local and global tools, but we need to be deliberate in what we support and do not support.

Reply:
cc @wli3

Reply:
Here was a long standing issue with the above feature - https://github.com/dotnet/cli/issues/2645. The issue was that this functionality didn't work when the project was a library (i.e. `netstandard`) because nothing was "runnable" and brought a runtime (neither the project nor the tool). Hopefully you can glean some answers from there.

My memory is a bit foggy, but I remember the tool itself would come through a dependency, and the CLI knew how to find it in the dependencies, and invoke it. Nate called it a `dependency command` or `"dependency"-type tool` in that issue.